featurebyte/_overrides/typechecked_override.py

```diff
@@ -0,0 +1,50 @@
+"""
+This module overrides the `typechecked` decorator from the `typeguard` package to customize
+type checking behavior based on the module where the function is defined.
+"""
+
+from typing import Any, Dict, Optional
+
+import functools
+import inspect
+
+import typeguard
+from typeguard import typechecked as original_typechecked
+
+
+def custom_typechecked(
+    func: Any = None,
+    *,
+    always: bool = False,
+    _localns: Optional[Dict[str, Any]] = None,
+) -> Any:
+    """
+    Customize the `typechecked` decorator to selectively enable or disable type checking
+    based on the defining module of the function or class.
+
+    Parameters
+    ----------
+    func : Any
+        The function or class to enable type checking for.
+    always : bool
+        True to enable type checks even in optimized mode.
+    _localns : Optional[Dict[str, Any]]
+        Local namespace used for resolving type annotations.
+
+    Returns
+    -------
+    Any
+        Either the decorated function or a partial object, depending on whether `func` is provided.
+    """
+    if func is None:
+        return functools.partial(custom_typechecked, always=always, _localns=_localns)
+
+    module = inspect.getmodule(func)
+    if module and module.__name__.startswith("feast."):
+        return func
```

Comment:
Does this completely disable the usage of typecheck everywhere in feast? Wondering if this will allow some errors to go unnoticed over time. Is there a more specific hotpath we can patch?

Reply:
Yes, this completely disable the usage of `typechecked` in feast. It seems to me that the `typechecked` is used to validate the user input if users use `feast` in an interactive way (or during development). The most significant one is the one used in `FeatureView` (https://github.com/gee-senbong/feast/blob/86d62215f2338ea9d48c6e723e907c82cbe5500b/sdk/python/feast/feature_view.py#L55). Should we just skip it for that module? Actually i was thinking whether to skip the check in https://github.com/featurebyte/featurebyte/blob/main/featurebyte/feast/patch.py#L248-L253.

Reply:
If the performance of skipping the check for a specific module is comparable, I think that is preferable. Otherwise I'm also fine with the current approach.

Reply:
Got it. IIRC, runtime of skipping the `FeatureView` class  should be comparable to skipping the whole package. Will run an experiment to double check.